### ISSUE TITLE
-O3 or -Ofast with -flto (gcc 9.2.0) fails to compile due to -ftree-loop-vectorize

### ISSUE THREAD

**Issue by OH-AU** [flame/blis#397]
blis 0.7.0
Although the above combination would not be used by many, in case it helps improve code in some way I thought I'd make the observation.
If I add -fno-tree-loop-vectorize to -O3 or -Ofast (with -flto enabled) the compile succeeds and check passes all tests. If I leave the option there I get the following error attempting to compile:
`kernels/haswell/3/sup/bli_gemmsup_rv_haswell_asm_d6x8n.c: In function ‘bli_dgemmsup_rv_haswell_asm_1x8n’:
kernels/haswell/3/sup/bli_gemmsup_rv_haswell_asm_d6x8n.c:3703:2: error: ‘asm’ operand has impossible constraints
 3703 |  begin_asm()`

Adding -ftree-loop-vectorize -O2 with -flto, the code compiles but fails cblat2 and zblat2 test (it passes without -ftree-loop-vectorize so it seem reasonably specific).
Other configure options were: --enable-blas --enable-cblas -t openmp haswell
CPU is an Intel Xeon E5-2660 v3.

**Comment by fgvanzee** [comment]
@OH-AU Thanks for letting us know about this. Just to clarify, how are you specifying the aforementioned options? For example, are you modifying the `make_defs.mk` file of the relevant subconfiguration for your hardware?

@devinamatthews Any ideas what's going on with @OH-AU 's observation?

**Comment by devinamatthews** [comment]
The "impossible constraints" error happens sometimes when the compiler reserves too many registers and not enough are left for the ASM block. Can happen when either too much debugging or too much optimization enabled.

@OH-AU what happens if you enable `-ftree-loop-vectorize` and `-flto` separately? In particular I am very wary of LTO. OTOH, cblat2 and zblat2 don't even touch any of the BLIS internals (they are straight f2c ports); perhaps they inherit some UB from Fortran?

**Comment by fgvanzee** [comment]
@devinamatthews Can you explain "UB from Fortran?"

**Comment by devinamatthews** [comment]
Fortran and C have different ideas about many things, for example aliasing. A straight port isn't guaranteed to adhere to strict C99 standards.

**Comment by OH-AU** [comment]
Also wary of LTO. And I'm just about to change the landscape with the below comments due to other flags used
Flags implemented via CFLAGS, CXXFLAGS and LDFLAGS - no manual modifications of any files.
"-O2 -fPIC -mtune=native -march=native -ftree-loop-vectorize" (ie - not LTO) results in the same blas errors on check - but compiles without error.
"-O2 -fPIC -ftree-loop-vectorize" - all tests pass. so looks like it doesn't like the native tune flags either.
When it fails tests, the output:
grep -i fail out.*
out.cblat2
55: ******* CTPMV  FAILED ON CALL NUMBER:
79: ******* CTPSV  FAILED ON CALL NUMBER:

out.zblat2
47: ******* ZTBMV  FAILED ON CALL NUMBER:
59: ******* ZTPMV  FAILED ON CALL NUMBER:
75: ******* ZTBSV  FAILED ON CALL NUMBER:
87: ******* ZTPSV  FAILED ON CALL NUMBER:
Basically, it looks like I've managed to force specific flags that are causing issues.

**Comment by devinamatthews** [comment]
@fgvanzee what flags do these files normally get (if configured for haswell)?

@OH-AU I believe we've had problems with these BLAS tests before, e.g. with newer Intel compilers. I blame the netlib source, but I guess the easy "workaround" is just to not use more aggressive compiler flags. The way BLIS is written, extra compiler flags like these shouldn't change performance much if at all.

**Comment by fgvanzee** [comment]
@devinamatthews Sorry, this fell through the cracks. Compiler flags for `haswell` vary based on which source files are being compiled.
- optimized kernels: `-O3 -Wall -Wno-unused-function -Wfatal-errors -fPIC -std=c99`
- reference kernels: `-O3 -mavx2 -mfma -mfpmath=sse -march=haswell -funsafe-math-optimizations -ffp-contract=fast -Wall -Wno-unused-function -Wfatal-errors -fPIC -std=c99`
- framework: `-O3 -Wall -Wno-unused-function -Wfatal-errors -fPIC -std=c99`

I agree with Devin that we've probably maxed out the utility of compiler flags in BLIS. Any new flags will either be of marginal benefit, or counterproductive.

**Comment by devinamatthews** [comment]
AFAICT `-flto` cannot handle different optimization flags in different object files. [This](https://gcc.gnu.org/wiki/lto/OptionHandling) seems to be some sort of working document to address that but as of now I think is is a no-no. I'm going to close this since it seems "Resolved".
